File: src/core/hardware.cpp
#include "core/hardware.hpp"
#include <cstdio>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

string getCpuModel() {
  ifstream cpuinfo("/proc/cpuinfo");
  string line;
  while (getline(cpuinfo, line)) {
    if (line.find("model name") != string::npos) {
      size_t colon = line.find(":");
      if (colon != string::npos) {
        return line.substr(colon + 2); // Skip ": "
      }
    }
  }
  return "Unknown CPU Model";
}

string getCpuCoreCount() {
  ifstream cpuinfo("/proc/cpuinfo");
  string line;
  int count = 0;
  while (getline(cpuinfo, line)) {
    if (line.find("processor") != string::npos) {
      count++;
    }
  }
  return to_string(count);
}

std::vector<std::string> getGpuModels() {
  std::string raw = execCmd("lspci | grep -Ei 'vga|3d|display'");
  std::vector<std::string> list;

  size_t start = 0;
  while (true) {
    size_t end = raw.find('\n', start);
    if (end == std::string::npos)
      break;
    std::string line = raw.substr(start, end - start);
    start = end + 1;

    size_t spacePos = line.find(' ');
    if (spacePos != std::string::npos)
      line = line.substr(spacePos + 1);

    size_t colonPos = line.find(": ");
    if (colonPos != std::string::npos)
      line = line.substr(colonPos + 2);

    list.push_back(line);
  }

  return list;
}

string getTotalRam() {
  ifstream meminfo("/proc/meminfo");
  string line;
  while (getline(meminfo, line)) {
    if (line.find("MemTotal:") == 0) {
      size_t start = line.find_first_of("0123456789");
      size_t end = line.find(" kB", start);
      if (start != string::npos && end != string::npos) {
        string memKbStr = line.substr(start, end - start);
        long memKb = stol(memKbStr);
        long memMb = memKb / 1024;
        return to_string(memMb) + " MB";
      }
    }
  }
  return "Unknown RAM";
}

string getAvailableRam() {
  ifstream meminfo("/proc/meminfo");
  string line;
  while (getline(meminfo, line)) {
    if (line.find("MemAvailable:") == 0) {
      size_t start = line.find_first_of("0123456789");
      size_t end = line.find(" kB", start);
      if (start != string::npos && end != string::npos) {
        string memKbStr = line.substr(start, end - start);
        long memKb = stol(memKbStr);
        long memMb = memKb / 1024;
        return to_string(memMb) + " MB";
      }
    }
  }
  return "Unknown Available RAM";
}

string getUsedRam() {
  string total = getTotalRam();
  string available = getAvailableRam();
  if (total == "Unknown RAM" || available == "Unknown Available RAM")
    return "Unknown Used RAM";
  long totalMb = stol(total.substr(0, total.find(" ")));
  long availableMb = stol(available.substr(0, available.find(" ")));
  long usedMb = totalMb - availableMb;
  return to_string(usedMb) + " MB";
}

string getTotalSwap() {
  ifstream meminfo("/proc/meminfo");
  string line;
  while (getline(meminfo, line)) {
    if (line.find("SwapTotal:") == 0) {
      size_t start = line.find_first_of("0123456789");
      size_t end = line.find(" kB", start);
      if (start != string::npos && end != string::npos) {
        string swapKbStr = line.substr(start, end - start);
        long swapKb = stol(swapKbStr);
        long swapMb = swapKb / 1024;
        return to_string(swapMb) + " MB";
      }
    }
  }
  return "Unknown Swap";
}

string getSwapFree() {
  ifstream meminfo("/proc/meminfo");
  string line;
  while (getline(meminfo, line)) {
    if (line.find("SwapFree:") == 0) {
      size_t start = line.find_first_of("0123456789");
      size_t end = line.find(" kB", start);
      if (start != string::npos && end != string::npos) {
        string swapKbStr = line.substr(start, end - start);
        long swapKb = stol(swapKbStr);
        long swapMb = swapKb / 1024;
        return to_string(swapMb) + " MB";
      }
    }
  }
  return "Unknown Swap Free";
}

string getDiskSize() {
  string cmd = "df -h  | awk 'NR==2 {print $2}'";
  string size = execCmd(cmd.c_str());
  return trim(size);
}

string getDiskUsed() {
  string cmd = "df -h  | awk 'NR==2 {print $3}'";
  string used = execCmd(cmd.c_str());
  return trim(used);
}

string getResolution() {
  string cmd = "xdpyinfo | grep dimensions | awk '{print $2}'";
  string res = execCmd(cmd.c_str());
  return trim(res);
}

File: src/core/hardware.rs


pub fn get_cpu_model() -> String {
  let cpuinfo = std::fs::read_to_string("/proc/cpuinfo").unwrap();
  for line in cpuinfo.lines() {
    if line.starts_with("model name") {
      let parts: Vec<&str> = line.split(':').collect();
      if parts.len() > 1 {
        return parts[1].trim().to_string();
      }
    }
  }
  "Unknown CPU".to_string()
}

pub fn get_cpu_core_count() -> usize {
  let cpuinfo = std::fs::read_to_string("/proc/cpuinfo").unwrap();
  let mut count = 0;
  for line in cpuinfo.lines() {
    if line.starts_with("processor") {
      count += 1;
    }
  }
  count
}

pub fn get_gpu_models() -> Vec<String> {
  let lspci_output = std::process::Command::new("lspci")
    .arg(" | grep -Ei 'vga|3d|display'")
    .output()
    .expect("Failed to execute lspci command");
  let output_str = String::from_utf8_lossy(&lspci_output.stdout);
  let mut gpus = Vec::new();
  for line in output_str.lines() {
    if line.contains("VGA compatible controller") || line.contains("3D controller") {
      let parts: Vec<&str> = line.split('"').collect();
      if parts.len() > 3 {
        gpus.push(parts[3].to_string());
      }
    }
  }
  gpus
}

pub fn get_total_ram() -> String {
  let meminfo = std::fs::read_to_string("/proc/meminfo").unwrap();
  for line in meminfo.lines() {
    if line.starts_with("MemTotal:") {
      let parts: Vec<&str> = line.split_whitespace().collect();
      if parts.len() > 1 {
        if let Ok(kb) = parts[1].parse::<u64>() {
          return (kb / 1024).to_string() + " MB"; // Convert to MB
        }
      }
    }
  }
  "Unknown RAM".to_string()
}

pub fn get_available_ram() -> String {
  let meminfo = std::fs::read_to_string("/proc/meminfo").unwrap();
  for line in meminfo.lines() {
    if line.starts_with("MemAvailable:") {
      let parts: Vec<&str> = line.split_whitespace().collect();
      if parts.len() > 1 {
        if let Ok(kb) = parts[1].parse::<u64>() {
          return (kb / 1024).to_string() + " MB"; // Convert to MB
        }
      }
    }
  }
  "Unknown Available RAM".to_string()
}

pub fn get_used_ram() -> String {
  let total_ram_str = get_total_ram();
  let available_ram_str = get_available_ram();

  let total_ram_mb: u64 = total_ram_str
    .split_whitespace()
    .next()
    .unwrap_or("0")
    .parse()
    .unwrap_or(0);
  let available_ram_mb: u64 = available_ram_str
    .split_whitespace()
    .next()
    .unwrap_or("0")
    .parse()
    .unwrap_or(0);

  if total_ram_mb >= available_ram_mb {
    let used_ram_mb = total_ram_mb - available_ram_mb;
    return used_ram_mb.to_string() + " MB";
  }

  "Unknown Used RAM".to_string()
}

pub fn get_total_swap() -> String {
  let meminfo = std::fs::read_to_string("/proc/meminfo").unwrap();
  for line in meminfo.lines() {
    if line.starts_with("SwapTotal:") {
      let parts: Vec<&str> = line.split_whitespace().collect();
      if parts.len() > 1 {
        if let Ok(kb) = parts[1].parse::<u64>() {
          return (kb / 1024).to_string() + " MB"; // Convert to MB
        }
      }
    }
  }
  "Unknown Swap".to_string()
}

pub fn get_swap_free() -> String {
  let meminfo = std::fs::read_to_string("/proc/meminfo").unwrap();
  for line in meminfo.lines() {
    if line.starts_with("SwapFree:") {
      let parts: Vec<&str> = line.split_whitespace().collect();
      if parts.len() > 1 {
        if let Ok(kb) = parts[1].parse::<u64>() {
          return (kb / 1024).to_string() + " MB"; // Convert to MB
        }
      }
    }
  }
  "Unknown Swap Free".to_string()
}

pub fn get_disk_size() -> String {
    let output = std::process::Command::new("df")
        .arg("-h")
        .output()
        .expect("Failed to execute df");

    let output_str = String::from_utf8_lossy(&output.stdout);

    if let Some(line) = output_str.lines().nth(1) {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() > 1 {
            return parts[1].to_string();
        }
    }

    "Unknown Disk Size".to_string()
}

pub fn get_disk_used() -> String {
    let output = std::process::Command::new("df")
        .arg("-h")
        .output()
        .expect("Failed to execute df");

    let output_str = String::from_utf8_lossy(&output.stdout);

    if let Some(line) = output_str.lines().nth(1) {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() > 2 {
            return parts[2].to_string();
        }
    }

    "Unknown Disk Used".to_string()
}

pub fn get_resolution() -> String {
    let output = std::process::Command::new("xdpyinfo")
        .output()
        .expect("Failed to execute xdpyinfo");

    let output_str = String::from_utf8_lossy(&output.stdout);

    for line in output_str.lines() {
        if line.contains("dimensions:") {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 {
                return parts[1].to_string();
            }
        }
    }

    "Unknown Resolution".to_string()
}
